### ISSUE TITLE
RTOS threads, where are ?

### ISSUE THREAD

**Issue by flexiti** [SmingHub/SmingRTOS#80]
I have not used yet RTOS version but ...
Is it possible to move applications without changes to the version of the RTOS?
ROTS has many advantages, but it also has drawbacks.
eg. Neopixel work which is very dependent on the time (in "Arduino").
By  me transmission can be interrupted by other threads (threads is the main advantage RTOS).
How was this resolved?
I do not see any example showing the effect of RTOS, simply acts as a non rtos.
Something skipped briefly reviewing?

**Comment by avr39-ripe** [comment]
I try to run my rather big app which works good (as good as I code it ;) ) on top of nonOS Sming. When I try to run it on top SmingRTOS it starts to behave odd - it crashes reboots and so on, @hreintke suggest to inspect code about stack usage, so I change app a bit, and it starts work bit stabler, but after about 1000-1200 seconds of rather normal work with plenty heap available internal webserver stops to respond on requests, but other app logic works fine, module can be reached by pings.. So IMHO porting app to RTOS is not such easy task as simple change several includes and recompile it.. We must find good practice to use RTOS and to port existing apps to rtos before we can use it in production.

**Comment by hreintke** [comment]
@flexiti 
When porting SmingNONOS to SminRTOS I took the approach of : 

Step 1 
Port to RTOS limiting the required updates to Sming & Applications. That minimizes the effort needed and gives a base for further development. -> done, all sming framework functionality is available.
Step 2 
We now see the first larger applications being converted to SmingRTOS showing more of the effects/freertos specific application requirements -> in progress
Step 3  
Incorporate/make use of more freertos features withing Sming. Providing framework access to these.
-> need to be started. Ides and prototype implementations welcome.

**Comment by flexiti** [comment]
@hreintke 
**At the beginning I would not transfer the applications that are dependent on the execution time.**

Theorizing - RTOS gives portions of time for each thread, this means that applications should take this into account.
I did not check (I will :))
 what type of this RTOS is here.
eg. Round-robin scheduling, Static time scheduling etc.
From this should come out the assumptions and limitations for the application.
**It must be clear to writing software.**
(As how long it may take one single execution thread, or you can block other threads, or its own thread could be interrupted - in this case, for example. Neopixel will not work well)
All these examples that are implemented operate if there is a single thread.
(So basically a copy of the non sming RTOS).
In my opinion for each example you need to add a second thread (eg. that performs some empty math and printf score eg. every 1 sec etc), and then it turns out that examples work correctly when second thread running in the background

**Comment by hreintke** [comment]
@flexiti : 
Indeed there is a design difference between the NONOS and RTOS version of the Espressif SDK.
Although in my opinion, in both NONOS and RTOS implementations there will be timing interference when high(er) priority tasks are active within the SDK (f.e wifi). I've seen discussions on using NMI interrupts to prevent this to activities which are very timing sensitive.
In that respect your statement `

> It must be clear to writing software.

is true for all (not only esp8266) application development. And for sure, no architecture will be useful for all applications/hardware elements.

If you can take the action to 

> I did not check (I will :))
> what type of this RTOS is here

and feedback the results to the community there will be more clarity for all "Framework users" and "Framework developers".
Updating (part of) the examples in such a way that the effect of threading is visible can then be done based on that information.
I then would also suggest that there is an example which shows (part of) the full freertos possibilities.

**Comment by flexiti** [comment]
In single-threaded sming, if you start something, the next (another) part of the program is executed at the end of your part of program (except for a few exceptions).
In typical RTOS execution of your part of the program will be interrupted by the programs from other threads at intervals.
I am a little busy recently, but since I promised , today I will read about ESP8266  RTOS :)

**Comment by flexiti** [comment]
@hreintke 
I've had thoughts about RTOS, prepare an example.
But...
Can You tell me how looks Your physical path where You keep your source files?
(I have a problem with include files)
eg. My is:
"C:\tools\SmingRTOS" and inside  I have "sming" and "samples"
and esspresif is standard CHERTS location

**Comment by hreintke** [comment]
The location of the sources is handled by a number of env variables : 
Looking at your text above I think it should be : 
- ESP_HOME = c:\espressif -> will find the toolchain
- SDK_BASE = c:/espressif/esp8266_RTOS_SDK (forward slashes, I don't know how to fix it to allow backward slash for windows) -> will find the SDK 
- SMING_HOME = c:\tools\SmingRTOS\sming -> will find the sming files.

The search paths for compiler and linker are set in the makefiles (much cleaned in the RTOS version).

For applications it is enough to #include "SmingCore.h" which will provide all, except libraries.

**Comment by flexiti** [comment]
I know all this rules ;)
My question  was how You have , i will check why i have problem with includes

**Comment by hreintke** [comment]
This also how I started. Only difference my SMING_HOME=c:\Userdata\RTOS_Project\sming

Since a couple of days I use the Espressif RTOS SDK 1.4.0 in c:\Userdata\ESP8266_RTOS_SDK. This update was without any other changes.

What issues do you have with the includes ?

**Comment by flexiti** [comment]
> make all 
> C+ sming/appinit/user_main.cpp
> sming/appinit/user_main.cpp:3:40: fatal error: sming/include/sming_config.h: No such file or directory
> compilation terminated.
> make: *** [build/sming/appinit/user_main.o] Error 1

But the file exist and path also looks good

> SMING_HOME = c:\tools\SmingRTOS\sming 

at the evening I will continue looking for the cause

**Comment by flexiti** [comment]
@hreintke 
Compilation sming work for me now, but I do not know why before the change did not work
Makefile:
> ifeq ($(OS),Windows_NT)
> 	UNAME := Windows
> 	SMING_HOME ?= c:\tools\SmingRTOS\sming 

I changed:  (I removed the question mark)
> 	SMING_HOME = c:\tools\SmingRTOS\sming    

ps. Latter I will remove these few issue comments  because they do not fit to #80 :)

**Comment by hreintke** [comment]
@flexiti : 
You must have defined the SMING_HOME somewhere to something else, presumably the Sming_NONOS. When converting to SmingRTOS I kept to use the SMING_HOME probably that was a bad idea. 
I do not face this as I have my env variables defined in the eclipse workspace and use differenet ones for NONOS and RTOS and then can switch without any issues.

**Comment by flexiti** [comment]
@hreintke 
I removed all unrelated content, please do so also to have only those associated with #80

**Comment by flexiti** [comment]
Sming RTOS is based on http://www.freertos.org
(So we can use the descriptions from this page)

ESP8266 specific:
When using xTaskCreate to create a task, the task stack range is [176, 512].

If an array whose length is over 60 bytes is used in a task, it is suggested that users use
os_malloc and os_free rather than local variable to allocate array. 
Large local variables could lead to task stack overflow.

The RTOS SDK takes some priorities. 
Priority of the pp task is 13; 
Priority of precise timer (ms) thread is 12;
Priority of the TCP/IP task is 10; 
Priority of the freeRTOS timer is 2;
Priority of the idle task is 0.  (lowest)
Users can use tasks with priorities from 1 to 9.

If used  timer callback should not occupy CPU more than 15ms.

Default ESP RTOS tick is set to 10 ms

**Comment by hreintke** [comment]
Indeed the freertos pages are relevant for the SmingRTOS implementation, as are the espressif remarks on their sdk.

We need to get an understanding on how the callback/delegate based architecture is implemented within these tasks and how to use the within Sming to our advantage.

My current idea is that we need some way to get the callbacks/delegates being executed in a separate task. I did already use that within the HardwareSerial implementation (and in the application I am developing myself). That would for example also "solve" the 15ms limit on a timer callback.

**Comment by hreintke** [comment]
@flexiti : 
I have also seen (not looked in detail) this wrapper : https://github.com/JudeBake/FreeRTOS_EC/

**Comment by flexiti** [comment]
@hreintke 

> My doubts (and sure open for all discussions) are however that this does not show anything Sming 

Can You give example what it is Sming specyfic ?
Sming works on events, RTOS works on separate task ..so, I do not know whether this is a good mix it

**Basic_blink example** - blink led on one task, it is event based solution - work as non RTOS sming  - so why use an RTOS version ? In this case  only increases the length of the program

**Basic_blink_RTOS** example based on task, one task makes one job, second makes second job.
(in this particular example, one thread flashing red LED, the second thread is flashing blue,
You can imagine that eg. a one thread measures the temperature, then sends it to a server, the second thread waiting for the RS command and turns on specific devices, etc.)

Everything depends on what we want to do,
- Sometimes a good choice one loop (like Arduino loop)
- If we get a lot of messages a good idea to work on events (like Sming)
- And when we perform different tasks, often independent of each other a good idea is choosing RTOS. (like few Arduino loop working at the same time)

Therefore, in my opinion examples should be more "RTOS" type  than Sming type .. otherwise it will be a copy of the non RTOS sming with redundant RTOS

**Comment by PTDreamer** [comment]
My 2 cents.
Most of the callback architecture used in sming could be refactored to use queues instead, making them more rtos friendly.

**Comment by skandalfo** [comment]
I, for one, am happy with Sming providing an idiomatic C++ abstraction over the Espressif SDK.

Sming interfaces are often way more consistent than the underlying ESP APIs, and I appreciate that a lot. It might make sense for it to eventually provide wrappers for RTOS tasks and synchronisation mechanisms.

I am familiar with callbacks and find the delegate abstraction comfortable to use.

I don't think moving everything to use queues just because is convenient. When you don't care too much about the context in which your reactive code executes, a callback is fine; you just write the code for it with little boilerplate needed.

When you do care, you can always make the callback put the data into a queue anyway.

**Comment by skandalfo** [comment]
To be clear, I was talking about the user interface rather than the internals of the API.

I am myself using a "soft interrupt" abstraction where the interrupt just wakes a task to process callbacks.

I'm doing this because my callbacks talk to I2C devices, and I don't want the interrupts to be disabled while the CPU is bit banging the bus.

But for a different thing I'm using the hardware timer and doing everything from within the ISR just because I cannot tolerate timing interference from the SDK's high priority tasks; using tasks here wouldn't work.

So, if you ask me, my preferred API would be delegate-based, with options to specify the execution context: in-ISR, or delegating to a task of priority X.

That could be handled by the delegate class itself: a subclass might be constructed with a reference to an executor task, and its initial callback might just push the callback itself and arguments into the executor task's queue, where the real target would be executed.

**Comment by PTDreamer** [comment]
> I, for one, am happy with Sming providing an idiomatic C++ abstraction over the Espressif SDK.

Me too.

> I am familiar with callbacks and find the delegate abstraction comfortable to use.

Sure

> I don't think moving everything to use queues just because is convenient. When you don't care too much about the context in which your reactive code executes, a callback is fine; you just write the code for it with little boilerplate needed.

That completely goes against the reasoning behind using an RTOS. If you want to use a callback mechanism throughout you just don't use an RTOS.

**Comment by hreintke** [comment]
@PTDreamer : @all 
I will reply also on the other remarks but first.

The move to RTOS was based on two elements: 
- To be able to support the upcoming ESP32 which will only get a RTOS SDK
- To be able to have the RTOS options available

**Comment by hreintke** [comment]
There is now also the possibility to uses "Queued callbacks" and "Queued interrupts" which make use of the RTOS queues & tasks.
